spec/unit/actions/collect_segment_clients_spec.rb

```diff
@@ -37,46 +51,112 @@
-    let(:segment) { double('segment') }
-    let(:project) { double('project') }
+    let(:client_project) { double(GoodData::Project) }
+    let(:segment_master_project) { double(GoodData::Project) }
+    let(:segment) do
+      { segment_id: 'premium_segment',
+        development_pid: 'dev_project',
+        master_name: 'master_1',
+        segment: segment_mock,
+        segment_master: segment_master_project }
+    end
     let(:ads_client) { double('ads_client') }
     let(:ads_response) do
       [
         {
-          master_project_id: 'master_project_id'
+          master_project_id: master_project_id
         }
       ]
     end
     let(:params) do
-      params = {
+      {
         gdc_gd_client: gdc_gd_client,
         synchronize: [{}],
         segments: [segment],
         domain: domain,
         ads_client: ads_client
       }
-      GoodData::LCM2.convert_to_smart_hash(params)
     end
 
     before do
-      allow(client).to receive(:project).and_return(project)
+      allow(client).to receive(:project).and_return(client_project)
       allow(client).to receive(:project?).and_return(true)
-      allow(gdc_gd_client).to receive(:projects).with('master_project_id').and_return(project)
       allow(client).to receive(:client_id).and_return('client_id')
-      allow(segment).to receive(:segment_id).and_return('segment-id')
+      allow(segment_mock).to receive(:segment_id).and_return('segment-id')
       allow(ads_client).to receive(:execute_select).and_return(ads_response)
-      allow(project).to receive(:pid)
-      allow(project).to receive(:title)
+      allow(client_project).to receive(:pid)
+      allow(client_project).to receive(:title)
+      module GoodData
+        class Datawarehouse
+          # to make the test work for Rubies other than JRuby
+        end
+      end
+      allow(ads_client).to receive(:class) { GoodData::Datawarehouse }
+      allow(segment_master_project).to receive(:pid)
     end
 
     it 'uses the project from client in segments' do
       expect(client).to receive(:project)
-      subject.class.call(params)
+      subject
+    end
+
+    context 'when there are multiple master project versions' do
+      let(:latest_master_project) { double(GoodData::Project) }
+      let(:latest_master_id) { 'latest_master_project_id' }
+      let(:ads_response) do
+        [
+          {
+            master_project_id: 'foo',
+            version: 1
+          },
+          {
+            master_project_id: latest_master_id,
+            version: 2
+          }
+        ]
+      end
+
+      before do
+        allow(gdc_gd_client).to receive(:projects)
+          .with(latest_master_id)
+          .and_return(latest_master_project)
+        allow(latest_master_project).to receive(:pid)
+        allow(latest_master_project).to receive(:title)
+      end
+
+      context 'when segment master is the same as latest master' do
+        let(:project_pid) { 'the same' }
+        before do
```

Comment:
Similar code found in domain_spec.rb:151, collect_segment_clients_spec.rb:134 (mass = 34)